### PR TITLE
Feat  xbar layout

### PR DESCRIPTION
**Pull request by theogainey**
## Description
Functionality to declaratively create xbar menu layouts was added in three commits. 
1. commit containing a set of utility functions for composing functions with pipes
2. commit reworking wordWrap to take an object instead of a string as a parameter 
3. commit adding a function that transform an xbar menu item into a formatted string, and a function that takes an array of xbar menu items and transforms each item, in the correct order, into a formatted string,  

## Spec
example what this pull request enables.  
 
```
const layout = xbar([
  {
    text: 'Quotes', 
  },
  separator,
  {
    text: jsonData.author,
    submenu: [
      {
        text: jsonData.quote,
        wordWrap: 40,
        size: 16,
        color: 'navy'
      }
    ],
  }
]);

//layout 
[
 'Quotes',
 '---',
 'author',
 '--quote text line 1 | size=16 | color=navy', 
 '--quote text line 2  | size=16 | color=navy',
]
```
See Issue: [FSA22V1-95](https://sparkbox.atlassian.net/browse/FSA22V1-95)

## To Validate

1. Make sure all PR Checks have passed.
2. Check that new tests reflect stated logic 
3. Check that code logic and naming is clear

### CODE REVIEW

File: src/format.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+const getFormattedOutputOptions = pipe(
+  ignoreText,
+  ignoreSubMenuLevel,
+  Object.entries,
+  formatOutputOptions,
+);
+
+//entry layer
+export const formatLine = (lineObj) =>
```
Comment (theogainey) [inline src/format.js:26]: The intent of this function is transform an object into a formatted string. This function is composed of 3 separate functions that break this work into parts. When reviewing please review if it is clear that the work is spilt into parts and that the intent of each of these parts is clear.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+//entry layer
+/**
+ * Prints output to xbar and returns printed output as an array of strings
+ * @param layout - array of menu items
+ * @returns {string[]} array of strings that were printed as menu items
+ */
+export const xbar = (layout) =>
+  layout.reduce((menuItems, item) => {
+    switch (true) {
```
Comment (theogainey) [inline src/index.js:45]: `switch (true) ` is used for control flow. When reviewing please review this for clarity.

File: src/accessibility.js [hunk 3]
```diff
@@ -1,3 +1,10 @@
+import {
+  pipe,
+  pipeCompatibleFilter,
+  pipeCompatibleMap,
+  pipeCompatibleReduce,
+  pipeCompatibleSplit,
+} from './utility.js';
```
Comment (catheraaine) [inline src/accessibility.js:7]: Good job using named imports/exports.

File: src/accessibility.js [hunk 4]
```diff
@@ -21,22 +28,39 @@ const reduceLines = (lineLength) =>
...
- * @param {number} lineLength - max length of a line
- * @returns {string[]}
- */
-export const wordWrap = (text, lineLength) => {
-  return text.split(' ')
-    .filter((e) => e !== '')
-    .reduce(reduceLines(lineLength), [''])
-    .map((e) => e.trim());
-};
+//word wrap transformations
```
Comment (catheraaine) [inline src/accessibility.js]: you should include a space between your `//` and the text of your comment 
```suggestion
// word wrap transformations
```

File: src/accessibility.js [hunk 5]
```diff
@@ -21,22 +28,39 @@ const reduceLines = (lineLength) =>
...
+const removeExtraSpaces = pipeCompatibleFilter((word) => word !== '');
+
+const reduceToLines = (maxLineLength) =>
+  pipeCompatibleReduce(reduceLines(maxLineLength))(['']);
+
+const passPropertiesToLInes = (propertiesToPass) =>
+  pipeCompatibleMap((line) => ({
+    text: line.trim(),
+    ...propertiesToPass,
+  }));
```
Comment (catheraaine) [inline src/accessibility.js:42]: I really like how you're naming these functions.

File: src/format.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import { pipe, pipeCompatibleReduce } from './utility.js';
+
+//line Object access layer
+const getText = (lineObj) => lineObj.text ?? '';
+const getSubMenuLevel = (lineObj) => lineObj.submenuLevel;
+// deno-lint-ignore no-unused-vars
+const ignoreSubMenuLevel = ({ submenuLevel, ...lineObj }) => lineObj;
+// deno-lint-ignore no-unused-vars
```
Comment (catheraaine) [inline src/format.js:8]: good work adding a linter!!

File: src/format.js (hunk 1 above)
Comment (catheraaine) [reply src/format.js:26]: Yeah, I believe it's clear. You've named your functions clearly. If you wanted to be extra explicit, you cloud also break this down into variables and then combine the variables in your return. 

```javascript
const subMenu = getSubMenuFormatting(lineObj);
const text = getFormattedText(lineObj);
const options = getFormattedOutputOptions(lineObj);
  
return (subMenu + text + options); 

```

The only thing I find unclear is `text`.

File: src/index.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+      case (isSeparator(item)):
```
Comment (catheraaine) [inline src/index.js:46]: maybe `needsSeparator`. I'm not sure why you would add a separator to a separator

File: src/utility.js [hunk 8]
```diff
@@ -0,0 +1,14 @@
+export const pipe = (...functions) =>
+  (x) =>
+    functions.reduce(
+      (acc, fn) => fn(acc),
+      x,
+    );
```
Comment (catheraaine) [inline src/utility.js:6]: I think I expect more brackets on this? That might be my linting settings though

File: test/accessibility.test.js [hunk 9]
```diff
@@ -2,44 +2,164 @@ import { assertEquals } from 'https://deno.land/std@0.128.0/testing/asserts.ts';
...
+    },
+    {
+      text: 'than science, Morty. A lot of',
+    },
+    {
+      text: 'people dont get that',
+    },
   ];
-  const actual = wordWrap(input, 30);
+  //const actual = wordWrap(input);
```
Comment (catheraaine) [inline test/accessibility.test.js]: Do you want this comment in here?

File: test/accessibility.test.js [hunk 10]
```diff
@@ -2,44 +2,164 @@ import { assertEquals } from 'https://deno.land/std@0.128.0/testing/asserts.ts';
...
+      text: 'people dont get that',
+      color: 'red',
+      size: 12,
+    },
+  ];
+  const actual = wordWrap(input);
   assertEquals(actual, expected);
 });
 
 Deno.test('wordWrap places words longer than max line length on their own line', () => {
```
Comment (catheraaine) [inline test/accessibility.test.js]: maybe 
'wordWrap splits phrases longer than max line length in to new lines' ?

File: test/accessibility.test.js [hunk 11]
```diff
@@ -2,44 +2,164 @@ import { assertEquals } from 'https://deno.land/std@0.128.0/testing/asserts.ts';
...
+      text: 'implementation',
+    },
+    {
+      text: 'is long',
+    },
+  ];
   assertEquals(result, expected);
 });
 
 Deno.test('wordWrap only includes one space between words', () => {
```
Comment (catheraaine) [inline test/accessibility.test.js:74]: This is such a good test case!!

File: test/index.test.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+} from 'https://deno.land/std@0.128.0/testing/asserts.ts';
+import { separator, xbar } from '../src/index.js';
+
+Deno.test('xbar returns separator as --- ', () => {
+  const input = [separator];
+  const actual = xbar(input);
+  assertEquals(actual, ['---']);
+});
+
+Deno.test('xbar returns an array of strings ', () => {
```
Comment (catheraaine) [inline test/index.test.js]: extra space in this one

File: test/index.test.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+      ],
+    },
+  ];
+  const actual = xbar(input);
+  actual.forEach((element) => {
+    assert(typeof element === 'string');
+  });
+});
+
+Deno.test('xbar handles an menu item with a submenu so that the submenu is returned before the next menu item ', () => {
```
Comment (catheraaine) [inline test/index.test.js]: extra space again

**Review by catheraaine — COMMENTED** [review]
This is excellent, @theogainey! I left some small comments, largely focused on syntax or naming. I'm especially impressed at the number of test cases you've included.

File: src/utility.js (hunk 8 above)
Comment (theogainey) [reply src/utility.js:6]: I found this function on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/reduce#function_composition_enabling_piping). Adding brackets causes it to break. Exactly why I am not exactly sure.

File: test/accessibility.test.js (hunk 10 above)
Comment (theogainey) [reply test/accessibility.test.js]: The description I currently have does not reflect the intent of that test. The intent was to prove that long words aren't getting broken into two lines. I changed it to 'wordWrap does not split words longer than the max line length into multiple lines'.

File: src/index.js (hunk 7 above)
Comment (theogainey) [reply src/index.js:46]: How does `includeSeparator` seem? The idea was when the function runs into a separator it adds it to the layout but you are correct that right now it does read like an additional separator is getting added

**Pull request merged**